### ISSUE TITLE
Dataclass "init" method should expect the type from a custom descriptor's "set" method

### ISSUE THREAD

**Issue by pappasam** [python/mypy#13856]
This issue is copied, mostly verbatim, from here: <https://github.com/microsoft/pyright/issues/3245>.

If a `@dataclass` field is initialized with a descriptor as a default value, mypy should use the value type of the `__set__` method for the descriptor when determining the type of the corresponding parameter within the synthesized `__init__` method.

Types are also evaluated incorrectly in this case.

```python
from dataclasses import dataclass
from typing import Any, cast


class MyDescriptor:
    def __get__(self, __obj: object | None, __owner: Any) -> "int | MyDescriptor":
        if __obj is None:
            return self
        return cast(Any, __obj)._x

    def __set__(self, __obj: object | None, __value: int) -> None:
        if __obj is not None:
            cast(Any, __obj)._x = __value


@dataclass
class Foo:
    y: MyDescriptor = MyDescriptor()


f = Foo(3)  # Currently a false positive error in mypy

print(f.y)  # 3
f.y = 4
print(f.y)  # 4
print(Foo.y)  # MyDescriptor object

reveal_type(f.y) # int | MyDescriptor, which is wrong
reveal_type(Foo.y) # int | MyDescriptor, which is wrong
```

**To Reproduce**

Run mypy against the above script.

**Expected Behavior**

No Errors

**Actual Behavior**

```
test.py:23: error: Argument 1 to "Foo" has incompatible type "int"; expected "MyDescriptor"
test.py:30: note: Revealed type is "Union[builtins.int, test.MyDescriptor]"
test.py:31: note: Revealed type is "Union[builtins.int, test.MyDescriptor]"
Found 1 error in 1 file (checked 1 source file)
```

**Your Environment** 

- OS: Ubuntu 22.04.1
- Mypy version used: 0.982
- Mypy command-line flags: None
- Mypy configuration options from `mypy.ini` (and other config files): None
- Python version used: 3.10.6

**Comment by pappasam** [comment]
@erictraut fyi, I migrated your issue from pyright here

**Comment by debonte** [comment]
FYI, the behavior of `dataclass` with descriptor-typed fields was discussed in https://github.com/python/cpython/issues/91330 and is documented [here](https://docs.python.org/3.11/library/dataclasses.html?highlight=dataclasses#descriptor-typed-fields).

**Comment by glyph** [comment]
I think this confusion goes a bit deeper than `__set__`; Mypy seems to have the wrong idea about dataclass-like types in general.  Consider this brief comparison of attrs and dataclasses:

```python3
from __future__ import annotations
from typing import reveal_type, TYPE_CHECKING
from attr import define

class DecoGet:
    pass

class Deco:
    def __set__(self, oself: object, value: Deco) -> None:
        ...
    def __get__(self, oself: object, owner: None | type[object]=None) -> DecoGet:
        return DecoGet()

@define
class AttrBlub:
    a: Deco
    b: Deco = Deco()


from dataclasses import dataclass

@dataclass
class DataBlub:
    a: Deco
    b: Deco = Deco()

reveal_type(AttrBlub(Deco()).a)
reveal_type(AttrBlub(Deco()).b)

reveal_type(DataBlub(Deco()).a)
reveal_type(DataBlub(Deco()).b)

reveal_type(AttrBlub.a)
reveal_type(AttrBlub.b)

try:
    reveal_type(DataBlub.a)     # AttributeError at runtime
except:
    print("skipped")
reveal_type(DataBlub.b)
```

Side-by-side, type-check and output looks like this (emojis added to indicate agreement / disagreement between type-time and runtime):

```
t.py:27: note: Revealed type is "t.DecoGet"    Runtime type is 'Deco'              ❌
t.py:28: note: Revealed type is "t.DecoGet"    Runtime type is 'Deco'              ❌
t.py:30: note: Revealed type is "t.Deco"       Runtime type is 'Deco'              ✅
t.py:31: note: Revealed type is "t.DecoGet"    Runtime type is 'DecoGet'           ✅
t.py:33: note: Revealed type is "t.DecoGet"    Runtime type is 'member_descriptor' ❌
t.py:34: note: Revealed type is "t.DecoGet"    Runtime type is 'member_descriptor' ❌
t.py:37: note: Revealed type is "t.DecoGet"    skipped                             ❌
t.py:40: note: Revealed type is "t.DecoGet"    Runtime type is 'DecoGet'           ✅
```

**Comment by glyph** [comment]
As I was playing with this I discovered a more specific problem in this neighborhood with Callable #14869

**Comment by pappasam** [comment]
Just tested out with `mypy 1.2` and this appears to have been resolved!
